python/pyspark/ml/tests/test_wrapper.py

```diff
@@ -28,6 +29,11 @@
 
 class JavaWrapperMemoryTests(SparkSessionTestCase):
 
+    @unittest.skipIf(
+        os.environ.get("PYSPARK_PIN_THREAD", "false").lower() == "true",
+        "When the thread is pinned between Python and JVM, detaching seems not "
+        "working. This is potentially a bug in Py4J and SPARK-18630 can be persistent when "
+        "PYSPARK_PIN_THREAD is enabled, skipping for now.")
```

Comment:
hmm, ok maybe take back my previous comment about thinking pinned mode is OK.  This sounds like a major problem.  Long-running job-server style deployments seem like cases where you'd want the advantage of this feature, but also where not having proper GC would be a huge problem.

Have you had a chance to look at this anymore, or ask the py4j community about it?

Reply:
Not yet. I think it might be a bug at `gateway.detach(java_obj)`. From my reading, it happens when the Java instance is copied via Py4J, which seems happening in ML; however, I don't think we don't do it often in other components. So I suspected the affected scope might be smaller than I think.

Anyway, I will have to double check the test and check along with Py4J community soon ..

Reply:
@squito, sorry for my late update. It seems passing when I added some sleeps. Seems it just needed some delays to avoid flakiness. I fixed the tests.